### PR TITLE
Form-submitting mobile workers invoicing

### PR DESCRIPTION
**Pull request by nospame**
## Product Description
_Will be_ visible on select customer invoices when the "form-submitting mobile worker" feature is added to an active software plan by ops.

## Technical Summary
[SAAS-15707](https://dimagi.atlassian.net/browse/SAAS-15707), [SAAS-15708](https://dimagi.atlassian.net/browse/SAAS-15708), [SAAS-15709](https://dimagi.atlassian.net/browse/SAAS-15709)
Adds a task to calculate form submitting mobile workers in all domains on the first of every month and store to the `FormSubmittingMobileWorkerHistory`, and a new `FormSubmittingMobileWorkerLineItemFactory` (verbose but descriptive/following existing pattern) that allows the feature to be invoiced correctly.

### Safety story
Manually tested in a local environment, as well as added automated tests. Mostly re-uses existing code, and will get QA on staging.

### Automated test coverage
Automated tests added in `corehq.apps.accounting.tests.test_invoicing` and `...test_domain_user_history`

### QA Plan
QA once [SAAS-15710](https://dimagi.atlassian.net/browse/SAAS-15710) is written and on staging, since that will help facilitate testing.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

[SAAS-15710]: https://dimagi.atlassian.net/browse/SAAS-15710?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[SAAS-15707]: https://dimagi.atlassian.net/browse/SAAS-15707?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SAAS-15708]: https://dimagi.atlassian.net/browse/SAAS-15708?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SAAS-15709]: https://dimagi.atlassian.net/browse/SAAS-15709?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: corehq/apps/accounting/invoicing.py [hunk 1]
```diff
@@ -723,6 +725,7 @@ def unit_cost(self):
 
 
 class UserLineItemFactory(FeatureLineItemFactory):
+    history_cls = DomainUserHistory
```
Comment (mjriley) [inline corehq/apps/accounting/invoicing.py:728]: It looks like `UserLineItemFactory` already had a child in `WebUserLineItemFactory`, so with a third class added, I'm wondering if it's time to think about the class hierarchy. It seems like a smell that `WebUserLineItemFactory` will inherit the `history_cls` attribute, yet will have no use for it. Do you think that's an issue?

**Review by mjriley — COMMENTED** [review]
The code looks solid, but I did leave a few questions

File: corehq/apps/accounting/tasks.py [hunk 2]
```diff
@@ -810,6 +812,44 @@ def calculate_users_in_all_domains(today=None):
...
+        FormES(for_export=True)
+        .domain(domain)
+        .submitted(gte=start, lt=end)
+        .user_type('mobile')
+        .user_aggregation()
+        .size(0)
+        .run()
+        .aggregations.user.normalized_buckets
+    )
+    return len(form_counts_by_worker)
```
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/accounting/tasks.py:829]: We still get ids of the forms mapped to each user (from `normalized_buckets`), but it avoids returning the documents themselves. It was modified  from this similar query: https://github.com/dimagi/commcare-hq/blob/bdfe26a9b6881d6801f39c23d6ce54011012c417/corehq/apps/data_analytics/esaccessors.py#L59-L66

Specifically by not fetching `user_ids` first because we can specify the `user_type`, and by just using `normalized_buckets` because `counts_by_bucket` does more work than we needed here. 

There might still be a nicer way to _just_ get the user ids or even the count of users directly, but I'm not aware of one.

File: corehq/apps/accounting/tasks.py [hunk 3]
```diff
@@ -810,6 +812,44 @@ def calculate_users_in_all_domains(today=None):
...
+    for domain in Domain.get_all_names():
+        num_workers = count_form_submitting_mobile_workers(domain, date_start, today)
+        record_date = today - relativedelta(days=1)
+        try:
+            FormSubmittingMobileWorkerHistory.objects.create(
+                domain=domain,
+                num_users=num_workers,
+                record_date=record_date
+            )
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/accounting/tasks.py:845]: I don't think we're expecting anything to go wrong, but because the object it creates is used directly for invoicing we want to log if anything ever does go wrong. Or at least, that's my understanding of it, since catching the generic `Exception` like this is all over accounting tasks.

In practice, I think the only thing that ever could go wrong is if there is already a record matching that date and domain, which raises `IntegrityError` for violating the uniqueness constraint. I can catch that here instead.

File: corehq/apps/accounting/tests/test_domain_user_history.py [hunk 4]
```diff
@@ -51,6 +56,60 @@ def test_calculate_users_in_all_domains(self):
...
+
+    def test_form_submitting_mobile_worker_history(self):
+        worker_history = FormSubmittingMobileWorkerHistory.objects.create(
+            domain=self.domain.name,
+            num_users=self.num_form_submitting_workers,
+            record_date=self.one_day_ago
+        )
+        self.assertEqual(worker_history.domain, self.domain.name)
+        self.assertEqual(worker_history.num_users, self.num_form_submitting_workers)
+        self.assertEqual(worker_history.record_date, self.one_day_ago)
```
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/accounting/tests/test_domain_user_history.py:101]: I think you're correct. I pulled this over from testing for the similar `DomainUserHistory` without much consideration.

File: corehq/apps/accounting/tasks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mjriley) [reply corehq/apps/accounting/tasks.py:845]: I hear you on a catch-all statement for top-level code that is the last point of interaction. But here, this doesn't appear like that, because it's nested inside the domain loop. Given the unreliability of network connections, we could easily see errors occurring during `Domain.get_all_names()` or `count_form_submitting_mobile_workers`, but those errors don't have any error handling. Because of that, it looks to me like we're trying to catch a specific situation with user creation, and in that event, I'd want to be explicit about what errors we're trying to catch

File: corehq/apps/accounting/tests/test_domain_user_history.py [hunk 5]
```diff
@@ -51,6 +56,60 @@ def test_calculate_users_in_all_domains(self):
...
+
+    def test_calculate_form_submitting_mobile_workers_in_all_domains(self):
+        tasks.calculate_form_submitting_mobile_workers_in_all_domains()
+        self.assertEqual(FormSubmittingMobileWorkerHistory.objects.count(), 1)
+
+        worker_history = FormSubmittingMobileWorkerHistory.objects.first()
+        self.assertEqual(worker_history.domain, self.domain.name)
+        self.assertEqual(worker_history.num_users, self.num_form_submitting_workers)
+        self.assertEqual(worker_history.record_date, self.one_day_ago)
+
```
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/accounting/tests/test_domain_user_history.py:111]: > what do you think of moving it to a `test_tasks.py` module instead?

I think that makes sense here. Since we can trust that Django creates an object using the properties we give it, what's left to be tested here is really just the task that's used to generate the `FormSubmittingMobileWorkerHistory`.

File: corehq/apps/accounting/tests/test_invoicing.py [hunk 6]
```diff
@@ -450,6 +454,107 @@ def test_community_over_limit(self):
...
+            feature__feature_type=FeatureType.FORM_SUBMITTING_MOBILE_WORKER)[:1].get()
+
+    def create_worker_history(self, history_cls, record_date, num_workers):
+        history_cls.objects.create(
+            domain=self.domain,
+            record_date=record_date,
+            num_users=num_workers
+        )
+
+    def test_under_limit(self):
```
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/accounting/tests/test_invoicing.py:491]: That's a really good point. I was following what had been done for testing `WebUserLineItemFactory`, but that at least has a chunk of different logic, where `FormSubmittingMobileWorkerLineItemFactory` really is just a couple lines of difference.

FWIW we _don't_ currently have `TestLineItemFactory` but we could. I would see that being where testing for `under_limit` and `over_limit` happens. The one other thing that would want to be tested, which currently gets looked at implicitly, is whether adding the right `Feature` on the `SoftwarePlanVersion` causes that feature to show up correctly on an invoice. I think those might make sense as asserts on `TestLineItemFactory`, something like `test_form_submitting_mobile_worker_feature_on_invoice`.

File: corehq/apps/accounting/invoicing.py [hunk 7]
```diff
@@ -794,6 +797,14 @@ def unit_description(self):
         return self._unit_description_by_user_type("mobile user")
 
 
+class FormSubmittingMobileWorkerLineItemFactory(UserLineItemFactory):
+    history_cls = FormSubmittingMobileWorkerHistory
+
+    @property
+    def unit_description(self):
+        return super()._unit_description_by_user_type("form-submitting mobile worker")
```
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/accounting/invoicing.py:805]: > 1. `UserLineItemFactory` uses "mobile user", whereas we are using "mobile worker" here. Are they referring to different things, or should we standardize?

"User" and "mobile worker" here are the same thing. I'll check in with ops and see what they think about naming on invoices. They also had some lack of clarity about where/when the distinction between "web _user_" and "mobile _worker_" came about.


> 2. This appears to be a problem with `UserLineItemFactory` as well, but `_unit_description_by_user_type` translates its strings. Because the `user_type` is only injected after translation, it looks like "form-submitting mobile worker" (and "mobile user") will not get translated. I'd imagine these strings should be translated.

I think you're probably right, but again I'll check with ops since changing it on existing features would apply across all translated invoices. Would that just be a matter of restructuring this `ngettext` block? https://github.com/dimagi/commcare-hq/blob/cb08c50fa38b142b2657b1e5bfcded89916c904c/corehq/apps/accounting/invoicing.py#L780-L789

File: corehq/apps/accounting/tests/test_invoicing.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/accounting/tests/test_invoicing.py:497]: Rather than rework a bulk of the invoicing test suite, I opted for what I felt like was a happy-medium. I restructured so what we're testing are mainly the things that are based on the Form-Submitting Mobile Worker `FeatureRate`, rather than every generic line item property, and then the `unit_description`. https://github.com/dimagi/commcare-hq/pull/34868/commits/0afcabb5d356eea7eb341995e07682aa97dd8d66

File: corehq/apps/accounting/invoicing.py (hunk 1 above)
Comment (nospame) [reply corehq/apps/accounting/invoicing.py]: After discussing yesterday I think this is basically the solution we talked about - removing the `history_cls` as a class attribute and putting the subclass-specific functionality into something that gets called from within `quantity`. https://github.com/dimagi/commcare-hq/pull/34868/commits/c929dde08be8fcacce5e1aa13b54c8cf5e3cb683

File: corehq/apps/accounting/tasks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/accounting/tasks.py]: I opted to catch the specific exception that would get called if the user history already exists based on the `unique_together` constraint -- which is what would happen if the task got interrupted and restarted partway through, the most plausible situation for this to handle. https://github.com/dimagi/commcare-hq/pull/34868/commits/d7120031b0ec9f029495f322e7977bf660f216f9

File: corehq/apps/accounting/tasks.py [hunk 8]
```diff
@@ -810,6 +812,44 @@ def calculate_users_in_all_domains(today=None):
...
+        .user_aggregation()
+        .size(0)
+        .run()
+        .aggregations.user.normalized_buckets
+    )
+    return len(form_counts_by_worker)
+
+
+@periodic_task(run_every=crontab(hour=1, minute=0, day_of_month='1'), acks_late=True)
+def calculate_form_submitting_mobile_workers_in_all_domains(today=None):
```
[the remark this answers is not in the document]
Comment (mjriley) [reply corehq/apps/accounting/tasks.py:833]: We did a sanity-check on this, pulling the most form-dense domain on staging and determining how long a similar elasticsearch query would take to aggregate by users -- it completed in milliseconds. While production does have a lot of domains, and we could probably optimize this query by batching several domains per elasticsearch call, I think this query is safe as is.

File: corehq/apps/accounting/tests/test_invoicing.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mjriley) [reply corehq/apps/accounting/tests/test_invoicing.py:497]: I do notice we still preserved the tests for over and under limits, while it sounds like we're mainly just trying to verify that the specific changes for form submitting workers is different. Do you think the benefit of maintaining these tests outweighs the duplication and complexity of tests that are, at least by their name, trying to verify other things? Would it be possible to replace these with a single test verifying the unique features instead?

File: corehq/apps/accounting/invoicing.py (hunk 1 above)
Comment (mjriley) [reply corehq/apps/accounting/invoicing.py]: Changes look good!

**Review by mjriley — APPROVED** [review]
Looks good. Approving, as I don't think the remaining items are blocking

File: corehq/apps/accounting/tasks.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/accounting/tasks.py:833]: So, exactly as written this runs in about 35 seconds on staging. I'm betting iterating through all domains and writing a `FormSubmittingMobileWorkerHistory` is the bottleneck here, since we found that the ES query for any given domain is near-instantaneous. If production has around an order of magnitude more data than staging, that puts this task at roughly 5 minutes to run.

The other option, I think, is to only write the `FormSubmittingMobileWorkerHistory` for domains on a subscription that uses that feature, by prefiltering the list of domains. Something like:
```python
    form_submitting_feature_rates = FeatureRate.objects.filter(
        feature__feature_type=FeatureType.FORM_SUBMITTING_MOBILE_WORKER
    )
    relevant_domains = Subscription.visible_objects.filter(
        plan_version__feature_rates__in=form_submitting_feature_rates
    ).values_list('subscriber__domain').distinct()

    for domain in relevant_domains:
```

**Comment by nospame** [comment]
So, the test errors coming up here show that, since those tests are implicitly calling `count_form_submitting_mobile_workers` which uses an Elasticsearch query, but ES seems not to get set up for those tests, it's trying to access an index that doesn't exist. This basically makes sense, but I'm so far unable to recreate the error in my local environment on any of the three test classes coming up with errors (`TestWebUserBulkUpload`, `TestMobileUserBulkUpload`, `TestUserBulkUploadStrongPassword`), all of which are in the same `test_importer` file.

I'm not entirely sure what's going on here. Based on the stack trace it seems like celery may be involved, which would make sense as a difference on my local environment.

File: corehq/apps/accounting/user_text.py [hunk 9]
```diff
@@ -54,6 +54,7 @@
     FeatureType.SMS: _("SMS Messages"),
     FeatureType.USER: _("Mobile Workers"),
     FeatureType.WEB_USER: _("Web Users"),
+    FeatureType.FORM_SUBMITTING_MOBILE_WORKER: _("Form-Submitting Mobile Workers"),
```
Comment (mjriley) [inline corehq/apps/accounting/user_text.py:57]: Just want to highlight that it's weird both `FEATURE_TYPE_TO_NAME` and `get_feature_name` exist. I'm not sure one translating 'SMS' to 'SMS Messages' while the other translates it to 'SMS' is a feature or a bug. Probably out of scope

File: corehq/apps/accounting/usage.py [hunk 10]
```diff
@@ -21,14 +21,19 @@ def __init__(self, feature_rate, domain_name,
...
             today.year, today.month, last_day)
 
+    @property
+    def usage_fns(self):
+        return {
+            FeatureType.USER: self._get_user_usage,
+            FeatureType.SMS: self._get_sms_usage,
+            FeatureType.WEB_USER: self._get_web_user_usage,
+            FeatureType.FORM_SUBMITTING_MOBILE_WORKER: self._get_form_submitting_mobile_worker_user_usage,
+        }
```
Comment (mjriley) [inline corehq/apps/accounting/usage.py:31]: It doesn't look like there is logic here, so it appears `usage_fns` could be a dictionary that is initialized in the constructor. What is the advantage of using a property here?

File: corehq/apps/accounting/invoicing.py [hunk 11]
```diff
@@ -797,7 +797,7 @@ def _unit_description_by_user_type(self, user_type):
 
     @property
     def unit_description(self):
-        return self._unit_description_by_user_type("mobile user")
+        return self._unit_description_by_user_type(_("mobile user"))
```
Comment (mjriley) [inline corehq/apps/accounting/invoicing.py:800]: Purely going off the method name, `_unit_description_by_user_type`, I'd assume this is taking a language-agnostic user type and intending to return a potentially translated description. Are we sure we want to translate the user type instead, returning whatever this funtion returns, untranslated?

**Review by mjriley — COMMENTED** [review]
Only pressing concern is the translation issue. Apart from that, looks good

File: corehq/apps/accounting/usage.py (hunk 10 above)
Comment (nospame) [reply corehq/apps/accounting/usage.py:31]: Do you mean like this?
```python
    def __init__(self, feature_rate, domain_name,
                 start_date=None, end_date=None):
        ...
        self.usage_fns = {
            ...
        }
```
I felt a property was more readable, but only by a small amount.

Comment (mjriley) [reply corehq/apps/accounting/usage.py:31]: Yeah. I guess I have a bias against properties

File: corehq/apps/accounting/invoicing.py (hunk 11 above)
Comment (nospame) [reply corehq/apps/accounting/invoicing.py:800]: This would be passing in the translated user type so it can be used in a string that itself will get translated [here](https://github.com/dimagi/commcare-hq/pull/34868/files#diff-d80a36dc3e18177c7a892bde9bc9aec4665cbd55d164e3a38b377ecf99a49bccR786-R796), however it is possible it will not get pluralized correctly (it never would have, though). Would it be more appropriate to instead build the singular and plural strings first, then use `ngettext`, something like this?
```python
            singular_string = (
                f"Per-{user_type} fee exceeding limit of {self.rate.monthly_limit} {user_type} "
                f"with plan above.{prorated_notice}"
            )
            plural_string = (
                f"Per-{user_type} fee exceeding limit of {self.rate.monthly_limit} {user_type}s "
                f"with plan above.{prorated_notice}"
            )
            return ngettext(singular_string, plural_string, self.rate.monthly_limit)
```

File: corehq/apps/accounting/user_text.py (hunk 9 above)
Comment (nospame) [reply corehq/apps/accounting/user_text.py:57]: Thanks, I thought it was weird too but didn't look into it. It seems like `get_feature_name` is just used by the "Feature Summary" on the Current Subscription page, where `FEATURE_TYPE_TO_NAME` is for the user-facing description of a Software Plan Version (which is used in a bunch of places). I'm inclined to just point `get_feature_name` to the `FEATURE_TYPE_TO_NAME` dict, seeing that the function is only used in one place.

File: corehq/apps/accounting/invoicing.py (hunk 11 above)
Comment (mjriley) [reply corehq/apps/accounting/invoicing.py:800]: This is looking quite heavy-weight for a warning string, to me. I wonder if there's a more generic way we could phrase this warning to prevent so much translation. Especially with the `user_type` passed in being already translated, I'd imagine many languages would have no way to make the translations look correct. Something like `per-user`, in English, might not even involve 'user' in another language. If we can't make it more generic, I think we'd be better off passing in the untranslated `user_type` and handling the translation in this method, instead. I think the gives us more flexibility, and I think its a bit surprising to realize that `user_type` is a translated parameter.

Comment (nospame) [reply corehq/apps/accounting/invoicing.py:800]: > for a warning string

It reads as a warning, but this is actually the line-item description shown on an invoice when there are charges based on a particular user type, so it is important that it be quite specific. What if we instead pluralize just the `user_type` text? I'd remove the `_()` calls from each `unit_description` and instead do something like this:

```python
            user = ngettext(user_type, f"{user_type}s", self.rate.monthly_limit)
            return _(
                f"Per-{user} fee exceeding limit of {self.rate.monthly_limit} {user} "
                f"with plan above.{prorated_notice}"
            )
```

Comment (mjriley) [reply corehq/apps/accounting/invoicing.py:800]: Your presentation today helped me see where this was used. It feels like it could be something like "{user_type}: number of users above threshold of {limit}", and that would cut down on the awkward translation corner cases, but I'll leave that up to you. The only change I'd like to see is moving the `user_type` translation inside this method, and then whatever you come up with is fine.

Comment (nospame) [reply corehq/apps/accounting/invoicing.py]: https://github.com/dimagi/commcare-hq/pull/34868/commits/c2776b7dd5e276d0ff6bb5cad6dde4bd6bdaa361

File: corehq/apps/accounting/user_text.py (hunk 9 above)
Comment (nospame) [reply corehq/apps/accounting/user_text.py:57]: https://github.com/dimagi/commcare-hq/pull/34868/commits/a5ae823f1907ca4a6da4ce4874c9ea20df452e4b

File: corehq/apps/accounting/invoicing.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/accounting/invoicing.py:811]: There remains a small debate over naming that I'm waiting to hear on - ops would prefer to see "form-submitting mobile _user_" for consistency, but otherwise I think this is addressed.

File: corehq/apps/accounting/invoicing.py [hunk 12]
```diff
@@ -777,41 +783,44 @@ def _unit_description_by_user_type(self, user_type):
...
-                "Per-{user} fee exceeding limit of {monthly_limit} {user}s "
-                "with plan above.{prorated_notice}",
-                self.rate.monthly_limit
-            ).format(
-                monthly_limit=self.rate.monthly_limit,
-                prorated_notice=prorated_notice,
-                user=user_type
+            return _(
+                f"Fee for each {user_type} exceeding the plan limit of {self.rate.monthly_limit}."
+                + prorated_notice
```
Comment (mjriley) [inline corehq/apps/accounting/invoicing.py:788]: I like the more concise wording, but now that we're passing in an untranslated `user_type`, I think we need to do the translation here:
```
return _('Fee for each {user_type} exceeding the plan limit of {monthly_limit}.').format(
  user_type=_(user_type), monthly_limit=self.rate.monthly_limit) + prorated_notice

Comment (nospame) [reply corehq/apps/accounting/invoicing.py:788]: I get so mixed up with how translations work. Does translating as an f-string just work for the text that's static in it, then?

**Review by mjriley — APPROVED** [review]
Thanks for the changes! LGTM
